contents/javascript/Javascript-promises.md

```diff
@@ -89,58 +85,117 @@ These situations correspond to the 3 states of Promises in Javascript.
 
-function update() {
-    return new Promise((resolve, reject) => {
-        var x = "I have not been updated";
-        setTimeout(function() {
-            x = "I have been updated";
-            resolve(x);
+```javascript
+function executeWithDelay(val) {
+    // return val after a short wait
+    return new Promise(resolve => {
+        setTimeout(function(){
+            resolve(val);
         }, 1000);
     });
 }
 
-update().then(result => console.log(result));
+/** note these funcs now return a promise **/
+function getdata() {
+    return executeWithDelay('some random data');
+}
+
+function filterdata(data, callback) {
+    return executeWithDelay(data.split(' '));
+}
 
+function main() {
+    return getData()
+        .then(data => filterData(data))
+        .then(filtered => console.log(filtered));
+}
+
+main();
 ```
 
-The first step of creating a Promise involves invoking the `Promise` constructor which takes in just 1 parameter known as the executor function. This is the function that will be executed by the Promise. This function has two parameters: resolve, reject. In our example, we only use the resolve function to indicate that our promise has executed successfully with the result `x`. The reject function may be used to indicate that some error has occured during execution. 
+Using Promises keep the code to be clean and easy to read, which makes it easy to learn what the code does, instead of having to trace through the code written with callbacks.
 
-After defining the function, we call it, and we call another function `.then()` upon the result. `.then()` is used to consume the result from a fulfilled Promise and act on it. In this case, we are using it to log our result. More importantly, it lets us chain promises to avoid the deep nesting known as callback hell.
+### The imperative style of promises
 
-Here's the callback hell code rewritten with Promises.
+The "promise style", which uses `.then()` to pass data from one function to the next is comparable to functional programming. The original promise is passed from one `.then()` to the other, and with each `.then()`, a new promise is returned for the next `.then()` to work on.
 
-```javascript
+That is unlike the iterative programming style which is considered to be a more intuitive way of writing code. As such, the `async` and `await` keywords are implemented to facilitate a iterative way of using promises.
 
-function update(numTimes) {
-    return new Promise((resolve, reject) => {
-        var x = "I have not been updated";
-        setTimeout(function() {
-            x = "I have been updated " + numTimes + " times.";
-            console.log(x);
-            resolve(numTimes);
-        }, 1000);
-    });
-}
+The `async` keyword ensures that the function returns a promise, and as for `await`, you can think of it as waiting for some promise to return the value before continuing. So back to the same example.
```

Comment:
Perhaps you can link back to the 3 states of Promises, which was mentioned at the top?